<!DOCTYPE html>
<!--[if lt IE 7]>      <html class="no-js lt-ie9 lt-ie8 lt-ie7"> <![endif]-->
<!--[if IE 7]>         <html class="no-js lt-ie9 lt-ie8"> <![endif]-->
<!--[if IE 8]>         <html class="no-js lt-ie9"> <![endif]-->
<!--[if gt IE 8]><!-->
<html class="no-js">
<!--<![endif]-->

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title></title>
    <meta name="description" content="">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" href="">
</head>

<body>
    <!--[if lt IE 7]>
            <p class="browsehappy">You are using an <strong>outdated</strong> browser. Please <a href="#">upgrade your browser</a> to improve your experience.</p>
        <![endif]-->

    <script>
        //for the synch code - RESOLVE
        let e = new Promise((resolve, reject) => {
            console.log("sync code,step by step from topm to bottom")
            resolve("resolved");
        });

        e
            .then(e => console.log("then", e))
            .catch(e => console.log("catch", e))
            .finally(() => console.log("finally"));

        //for the synch code - REJECT
        let e1 = new Promise((resolve, reject) => {
            console.log("SYNC CODE, FROM TOP TO BOTTOM, STER BY STEP")
            reject("rejected");
        });

        e1
            .then(e1 => console.log("then", e1))
            .catch(e1 => console.log("catch", e1))
            .finally(() => console.log("finally"));





        //for the asynch code - RESOLVE
        let r = new Promise((resolve, reject) => {
            console.log("Will this come sooner?");
            setTimeout(() => {
                resolve("resolved, after 3 seconds");
            }, 3000);
        });

        r
            .then(r => console.log("then", r))
            .catch(r => console.log("catch", r))
            .finally(() => console.log("finally"));

        //for the asynch code - REJECT
        let r1 = new Promise((resolve, reject) => {
            console.log("executed sooner");
            setTimeout(() => {
                reject("rejected - 3 sec delay");
            }, 3000);
        });

        r1
            .then(r1 => console.log("then", r1))
            .catch(r1 => console.log("catch", r1))
            .finally(() => console.log("finally"));





        //many then's and the catch, finally, on (a) method in line 86, we use many .then() is called promise chaining
        let a = new Promise((resolve, reject) => {
            console.log("CONSOLE .THEN MULTIPLE TIMES");
            setTimeout(() => {
                reject("RESOLVED AFTER 3 SEC");
            }, 3000);
        });
        a
            .then(a => console.log("then", a)) // multiple times we can call promise using .then()
            .then(a => console.log("then", a)) // UNDEFIEND, BECAUSE WE ARE NOT CALLING RESOLVE CASE, check line 81 - 85, only reject is avail, also for RESOLVE() it will not work, cause it will work on diff promise(eg : PROISE HELL)
            .then(a => console.log("then", a))
            .then(a => console.log("then", a))
            .catch(a => console.log("catch", a)) // note only 1 catch() will work so, NO MULTIPLE .catch()
            .catch(a => console.log("catch", a))
            .then(a => console.log("then-CATCH", a))
            .then(a => console.log("then-CATCH", a))
            .then(a => console.log("then-CATCH", a))
            .finally(() => console.log("finally")) // multiple times we can call promise using .finally()
            .finally(() => console.log("finally"));






        // this is to return data of multiple of 2
        let v = 2;
        new Promise((resolve, reject) => {
                setTimeout(() => resolve(2 * 2), 2000);
            })
            .then(x => {
                console.log(x);
                return new Promise((resolve, reject) => {
                    setTimeout(() => resolve(x * 2), 2000);
                });
            })
            .then(x => {
                console.log(x);
                return new Promise((resolve, reject) => {
                    setTimeout(() => resolve(x * 2), 2000);
                });
            })
            .then(x => {
                console.log(x);
                return new Promise((resolve, reject) => {
                    setTimeout(() => resolve(x * 2), 2000);
                });
            })
            .finally(() => console.log("Finally it is done"));

        // this is to return data of multiple of 2, without promise - TRY IN CONSOLE
        let v1 = 2;
        new Promise((resolve, reject) => {
                setTimeout(() => resolve(2 * 2), 2000);
            })
            .then(x1 => {
                console.log(x1);
                return (x1 * 2);
            })
            .then(x1 => {
                console.log(x1);
                return (x1 * 5);
            })
            .then(x1 => {
                console.log(x1);
                return new Promise((resolve, reject) => {
                    setTimeout(() => resolve(x1 * 2), 2000);
                });
            })
            .finally(() => console.log("Finally it is done"));
    </script>
    </script>
</body>

</html>